pbj-core/pbj-compiler/src/main/java/com/hedera/pbj/compiler/impl/Common.java

```diff
@@ -172,14 +174,367 @@ public static String javaPrimitiveToObjectType(String primitiveFieldType) {
+					}
+					""").replace("$fieldName", f.nameCamelFirstLower());
+				}
+				else {
+					throw new RuntimeException("Unexpected field type for getting HashCode - " + f.type().toString());
+				}
+			}
+		}
+		return generatedCodeSoFar.indent(DEFAULT_INDENT * 2);
+	}
+
+	/**
+	 * Get the hashcode codegen for a optional field.
+	 * @param generatedCodeSoFar The string that the codegen is generated into.
+	 * @param f The field for which to generate the hash code.
+	 * @return Updated codegen string.
+	 */
+	@NonNull
+	private static String getOptionalHashCodeGeneration(String generatedCodeSoFar, Field f) {
+		switch (f.messageType()) {
+			case "StringValue" -> generatedCodeSoFar += (
+     			"""
+				if ($fieldName != DEFAULT.$fieldName) {
+					result = 31 * result + $fieldName.hashCode();
+				}
+				""").replace("$fieldName", f.nameCamelFirstLower());
+			case "BoolValue" -> generatedCodeSoFar += (
+				"""
+				if ($fieldName != DEFAULT.$fieldName) {
+					result = 31 * result + Boolean.hashCode($fieldName);
+				}
+				""").replace("$fieldName", f.nameCamelFirstLower());
+			case "Int32Value", "UInt32Value" -> generatedCodeSoFar += (
+     			"""
+				if ($fieldName != DEFAULT.$fieldName) {
+					result = 31 * result + Integer.hashCode($fieldName);
+				}
+				""").replace("$fieldName", f.nameCamelFirstLower());
+			case "Int64Value", "UInt64Value" -> generatedCodeSoFar += (
+     			"""
+				if ($fieldName != DEFAULT.$fieldName) {
+					result = 31 * result + Long.hashCode($fieldName);
+				}
+				""").replace("$fieldName", f.nameCamelFirstLower());
+			case "FloatValue" -> generatedCodeSoFar += (
+     			"""
+				if ($fieldName != DEFAULT.$fieldName) {
+					result = 31 * result + Float.hashCode($fieldName);
+				}
+				""").replace("$fieldName", f.nameCamelFirstLower());
+			case "DoubleValue" -> generatedCodeSoFar += (
+     			"""
+				if ($fieldName != DEFAULT.$fieldName) {
+					result = 31 * result + Double.hashCode($fieldName);
+				}
+				""").replace("$fieldName", f.nameCamelFirstLower());
+			case "BytesValue" -> generatedCodeSoFar += (
+     			"""
+				if ($fieldName != DEFAULT.$fieldName) {
+					result = 31 * result + ($fieldName == null ? 0 : $fieldName.hashCode());
+				}
+				""").replace("$fieldName", f.nameCamelFirstLower());
+			default -> throw new UnsupportedOperationException("Unhandled optional message type:" + f.messageType());
+		}
+		return generatedCodeSoFar;
+	}
+
+	/**
+	 * Get the hashcode codegen for a repeated field.
+	 * @param generatedCodeSoFar The string that the codegen is generated into.
+	 * @param f The field for which to generate the hash code.
+	 * @return Updated codegen string.
+	 */
+	@NonNull
+	private static String getRepeatedHashCodeGeneration(String generatedCodeSoFar, Field f) {
+		generatedCodeSoFar += (
+           """
+			java.util.List list$$fieldName = $fieldName;
+			for (Object o : list$$fieldName) {
+				if ($fieldName != DEFAULT.$fieldName) {
```

Comment:
This one is questionable. I would assume that for repeated fields, it makes sense to take all values into consideration, even default values

Reply:
Another thought... If a repeated field is empty, then it should probably be ignored

Reply:
I addressed both comments.